File: windows/WifiLink.cs
using System.Collections.Concurrent;
using System.Net;
using System.Net.NetworkInformation;
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace Remco;

/// <summary>
/// Wi-Fi link (laptop hotspot, phone hotspot or any shared network).
///
/// To avoid Windows Firewall prompts entirely, the PC never listens:
///   1. The PC broadcasts a small UDP "beacon" every second (outbound).
///   2. The phone answers with proof that it knows this PC's PIN
///      (Windows allows unicast answers to our own broadcasts).
///   3. The PC then opens an outbound TCP connection to the phone.
/// After that it's the same one-JSON-per-line protocol as Bluetooth.
/// </summary>
internal sealed class WifiLink : IDisposable
{
    public const int BeaconPort = 47801;

    public event Action<string>? Log;
    public event Action<string>? LineReceived;
    public event Action<bool, string>? ConnectionChanged;

    private readonly Settings _settings;
    public string BtAddress { get; set; } = "";

    private UdpClient? _udp;
    private volatile bool _running;
    private TcpClient? _client;
    private BlockingCollection<string>? _outbox;
    private int _gen;
    private int _connecting;
    private readonly LinkedList<string> _nonces = new();
    private readonly object _lock = new();

    public bool IsConnected => _client != null;
    public string PeerName { get; private set; } = "";

    public WifiLink(Settings settings) { _settings = settings; }

    public void Start()
    {
        _udp = new UdpClient(new IPEndPoint(IPAddress.Any, 0)) { EnableBroadcast = true };
        _running = true;
        new Thread(BeaconLoop) { IsBackground = true, Name = "wifi-beacon" }.Start();
        new Thread(ReceiveLoop) { IsBackground = true, Name = "wifi-replies" }.Start();
        Log?.Invoke("Wi-Fi: looking for the phone on " + string.Join(", ", LocalAddresses().Select(a => a.ToString())));
    }

    /// <summary>IPv4 addresses of this PC (for display).</summary>
    public static List<IPAddress> LocalAddresses()
    {
        var list = new List<IPAddress>();
        try
        {
            foreach (var ni in NetworkInterface.GetAllNetworkInterfaces())
            {
                if (ni.OperationalStatus != OperationalStatus.Up || ni.NetworkInterfaceType == NetworkInterfaceType.Loopback) continue;
                foreach (var ua in ni.GetIPProperties().UnicastAddresses)
                    if (ua.Address.AddressFamily == AddressFamily.InterNetwork) list.Add(ua.Address);
            }
        }
        catch { }
        return list;
    }

    private static List<IPAddress> BroadcastTargets()
    {
        var list = new List<IPAddress> { IPAddress.Broadcast };
        try
        {
            foreach (var ni in NetworkInterface.GetAllNetworkInterfaces())
            {
                if (ni.OperationalStatus != OperationalStatus.Up || ni.NetworkInterfaceType == NetworkInterfaceType.Loopback) continue;
                foreach (var ua in ni.GetIPProperties().UnicastAddresses)
                {
                    if (ua.Address.AddressFamily != AddressFamily.InterNetwork || ua.IPv4Mask == null) continue;
                    byte[] ip = ua.Address.GetAddressBytes(), mask = ua.IPv4Mask.GetAddressBytes();
                    if (mask.All(b => b == 0)) continue;
                    // On a phone hotspot the phone is this PC's gateway — beacon it directly too.
                    foreach (var gw in ni.GetIPProperties().GatewayAddresses)
                        if (gw.Address.AddressFamily == AddressFamily.InterNetwork && !gw.Address.Equals(IPAddress.Any)) list.Add(gw.Address);
                    var bc = new byte[4];
                    for (int i = 0; i < 4; i++) bc[i] = (byte)(ip[i] | ~mask[i]);
                    list.Add(new IPAddress(bc));
                }
            }
        }
        catch { }
        return list;
    }

    private void BeaconLoop()
    {
        while (_running)
        {
            try
            {
                string nonce = Convert.ToHexString(RandomNumberGenerator.GetBytes(6)).ToLowerInvariant();
                lock (_lock)
                {
                    _nonces.AddFirst(nonce);
                    while (_nonces.Count > 12) _nonces.RemoveLast();
                }
                var beacon = JsonSerializer.Serialize(new
                {
                    remco = 1, type = "beacon", id = _settings.Id, name = Environment.MachineName,
                    bt = BtAddress, nonce, connected = IsConnected
                });
                byte[] data = Encoding.UTF8.GetBytes(beacon);
                var targets = BroadcastTargets();
                if (IPAddress.TryParse(_settings.PhoneIp, out var manual)) targets.Add(manual);
                foreach (var t in targets.Distinct())
                {
                    try { _udp?.Send(data, data.Length, new IPEndPoint(t, BeaconPort)); } catch { }
                }
            }
            catch { }
            Thread.Sleep(1000);
        }
    }

    private void ReceiveLoop()
    {
        while (_running)
        {
            try
            {
                IPEndPoint? ep = null;
                byte[] data = _udp!.Receive(ref ep);
                if (ep == null) continue;
                using var doc = JsonDocument.Parse(data);
                var m = doc.RootElement;
                if (Str(m, "type") != "reply" || Str(m, "id") != _settings.Id) continue;
                string proof = Str(m, "proof");
                bool ok;
                lock (_lock) ok = _nonces.Any(n => Proof(_settings.Pin, n) == proof);
                if (!ok) continue;
                if (IsConnected || Interlocked.CompareExchange(ref _connecting, 1, 0) != 0) continue;
                int port = m.TryGetProperty("port", out var pv) && pv.TryGetInt32(out int p) ? p : 47802;
                string phone = Str(m, "phone");
                string pnonce = Str(m, "pnonce");
                var addr = ep.Address;
                new Thread(() => Connect(addr, port, phone, pnonce)) { IsBackground = true }.Start();
            }
            catch (ObjectDisposedException) { break; }
            catch (SocketException) { if (!_running) break; }
            catch { }
        }
    }

    public static string Proof(string pin, string nonce)
    {
        byte[] h = SHA256.HashData(Encoding.UTF8.GetBytes(pin + ":" + nonce));
        return Convert.ToHexString(h).ToLowerInvariant()[..16];
    }

    private void Connect(IPAddress addr, int port, string phone, string pnonce)
    {
        TcpClient? c = null;
        try
        {
            c = new TcpClient { NoDelay = true };
            var task = c.ConnectAsync(addr, port);
            if (!task.Wait(3000) || !c.Connected) { c.Dispose(); return; }
            c.ReceiveTimeout = 8000;  // phone pings every 2 s; silence = Wi-Fi lost
            var stream = c.GetStream();
            var hello = JsonSerializer.Serialize(new
            {
                e = "pc_hello", id = _settings.Id, name = Environment.MachineName, bt = BtAddress,
                proof = Proof(_settings.Pin, pnonce)
            }) + "\n";
            byte[] hb = Encoding.UTF8.GetBytes(hello);
            stream.Write(hb, 0, hb.Length);
            stream.Flush();
            Session(c, stream, $"{(phone.Length > 0 ? phone : "phone")} ({addr})");
            c = null; // owned by Session now
        }
        catch (Exception ex) { Log?.Invoke("Wi-Fi connect failed: " + ex.Message); }
        finally
        {
            c?.Dispose();
            Interlocked.Exchange(ref _connecting, 0);
        }
    }

    private void Session(TcpClient c, NetworkStream stream, string name)
    {
        int gen = Interlocked.Increment(ref _gen);
        var outbox = new BlockingCollection<string>(new ConcurrentQueue<string>());
        _outbox = outbox;
        _client = c;
        PeerName = name;
        ConnectionChanged?.Invoke(true, name);
        Log?.Invoke("Phone connected over Wi-Fi: " + name);

        var writer = new Thread(() =>
        {
            try
            {
                foreach (var line in outbox.GetConsumingEnumerable())
                {
                    byte[] b = Encoding.UTF8.GetBytes(line + "\n");
                    stream.Write(b, 0, b.Length);
                }
            }
            catch { }
        }) { IsBackground = true, Name = "wifi-writer" };
        writer.Start();

        // Reader runs on a background thread so the caller (Connect) returns.
        new Thread(() =>
        {
            try
            {
                using var reader = new StreamReader(stream, Encoding.UTF8);
                while (gen == _gen)
                {
                    string? line = reader.ReadLine();
                    if (line == null) break;
                    if (line.Length > 0) LineReceived?.Invoke(line);
                }
            }
            catch { }
            if (gen == _gen)
            {
                Drop();
                ConnectionChanged?.Invoke(false, name);
                Log?.Invoke("Wi-Fi connection closed. Waiting for the phone…");
            }
            try { c.Dispose(); } catch { }
        }) { IsBackground = true, Name = "wifi-reader" }.Start();
    }

    public void Send(string json)
    {
        var box = _outbox;
        if (box == null || box.IsAddingCompleted) return;
        try { box.Add(json); } catch (InvalidOperationException) { }
    }

    private void Drop()
    {
        Interlocked.Increment(ref _gen);
        try { _outbox?.CompleteAdding(); } catch { }
        _outbox = null;
        try { _client?.Dispose(); } catch { }
        _client = null;
    }

    private static string Str(JsonElement m, string k) =>
        m.TryGetProperty(k, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() ?? "" : "";

    public void Dispose()
    {
        _running = false;
        Drop();
        try { _udp?.Dispose(); } catch { }
    }
}

File: mac/lib/wifilink.js
// Wi-Fi link — same protocol as the Windows version (the phone app needs no change):
//  1. the Mac broadcasts a UDP beacon every second (port 47801),
//  2. the phone answers with a proof that it knows the PIN,
//  3. the Mac connects OUT to the phone over TCP and they talk one JSON object per line.
// macOS never has to accept incoming connections, so its firewall never asks.
const dgram = require('dgram'), net = require('net'), os = require('os'), crypto = require('crypto');
const { EventEmitter } = require('events');

const BEACON_PORT = 47801;
const proof = (pin, nonce) => crypto.createHash('sha256').update(`${pin}:${nonce}`).digest('hex').slice(0, 16);

function localAddresses() {
  const out = [];
  for (const list of Object.values(os.networkInterfaces()))
    for (const a of list || []) if (a.family === 'IPv4' && !a.internal) out.push(a);
  return out;
}

function broadcastTargets() {
  const t = new Set(['255.255.255.255']);
  for (const a of localAddresses()) {
    const ip = a.address.split('.').map(Number), mask = a.netmask.split('.').map(Number);
    t.add(ip.map((b, i) => (b | (~mask[i] & 255)) & 255).join('.'));
    // on a phone hotspot the phone is the gateway — usually x.x.x.1 of that network
    t.add(ip.map((b, i) => (b & mask[i])).map((b, i) => (i === 3 ? b + 1 : b)).join('.'));
  }
  return [...t];
}

class WifiLink extends EventEmitter {
  constructor(settings) {
    super();
    this.settings = settings;
    this.nonces = [];
    this.sock = null;       // TCP socket to the phone
    this.connecting = false;
    this.peer = '';
  }
  get connected() { return !!this.sock; }

  start() {
    this.udp = dgram.createSocket({ type: 'udp4', reuseAddr: true });
    this.udp.on('error', e => this.emit('log', 'Wi-Fi: ' + e.message));
    this.udp.on('message', (buf, rinfo) => this.onReply(buf, rinfo));
    this.udp.bind(0, () => {
      try { this.udp.setBroadcast(true); } catch { }
      this.timer = setInterval(() => this.beacon(), 1000);
      this.beacon();
    });
    this.emit('log', 'Wi-Fi: looking for the phone on ' + (localAddresses().map(a => a.address).join(', ') || 'no network'));
  }

  beacon() {
    const nonce = crypto.randomBytes(6).toString('hex');
    this.nonces.unshift(nonce); this.nonces.length = Math.min(this.nonces.length, 12);
    const msg = Buffer.from(JSON.stringify({
      remco: 1, type: 'beacon', id: this.settings.id, name: os.hostname().replace(/\.local$/, ''),
      bt: '', nonce, connected: this.connected
    }));
    const targets = broadcastTargets();
    if (this.settings.phoneIp) targets.push(this.settings.phoneIp);
    for (const t of targets) { try { this.udp.send(msg, BEACON_PORT, t); } catch { } }
  }

  onReply(buf, rinfo) {
    let m; try { m = JSON.parse(buf.toString('utf8')); } catch { return; }
    if (m.type !== 'reply' || m.id !== this.settings.id) return;
    if (!this.nonces.some(n => proof(this.settings.pin, n) === m.proof)) return;   // wrong PIN
    if (this.connected || this.connecting) return;
    this.connecting = true;
    const port = Number(m.port) || 47802;
    const s = net.connect({ host: rinfo.address, port, timeout: 3000 });
    s.setNoDelay(true);
    s.once('connect', () => {
      s.setTimeout(8000);               // the phone pings every 2 s; silence = Wi-Fi lost
      s.write(JSON.stringify({ e: 'pc_hello', id: this.settings.id, name: os.hostname(), bt: '', proof: proof(this.settings.pin, m.pnonce || '') }) + '\n');
      this.sock = s; this.connecting = false;
      this.peer = `${m.phone || 'phone'} (${rinfo.address})`;
      this.emit('connected', this.peer);
      this.emit('log', 'Phone connected over Wi-Fi: ' + this.peer);
    });
    let rest = '';
    s.on('data', d => {
      rest += d.toString('utf8');
      let i;
      while ((i = rest.indexOf('\n')) >= 0) {
        const line = rest.slice(0, i).trim(); rest = rest.slice(i + 1);
        if (line) this.emit('line', line);
      }
    });
    const drop = () => {
      this.connecting = false;
      if (this.sock === s) { this.sock = null; this.emit('disconnected'); this.emit('log', 'Wi-Fi connection closed. Waiting for the phone…'); }
      try { s.destroy(); } catch { }
    };
    s.on('timeout', drop); s.on('error', drop); s.on('close', drop);
  }

  send(obj) {
    if (!this.sock) return;
    try { this.sock.write(JSON.stringify(obj) + '\n'); } catch { }
  }

  stop() {
    clearInterval(this.timer);
    try { this.udp.close(); } catch { }
    try { this.sock && this.sock.destroy(); } catch { }
  }
}
module.exports = { WifiLink, localAddresses, proof };
